lf-queue/lfq.c

```diff
@@ -0,0 +1,223 @@
+    ctx->head = ctx->tail = tmp;
+    ctx->fph = ctx->fpt = node;
+
+    return 0;
+}
+
+long lfg_count_freelist(const struct lfq_ctx *ctx)
+{
+    long count = 0;
+    for (struct lfq_node *p = (struct lfq_node *) ctx->fph; p; p = p->free_next)
+        count++;
+    return count;
+}
+
+int lfq_release(struct lfq_ctx *ctx)
+{
+    if (ctx->tail && ctx->head) {               /* if we have data in queue */
+        while ((struct lfq_node *) ctx->head) { /* while still have node */
+            struct lfq_node *tmp = (struct lfq_node *) ctx->head->next;
+            safe_free(ctx, (struct lfq_node *) ctx->head);
+            ctx->head = tmp;
+        }
+        ctx->tail = 0;
+    }
+    if (ctx->fph && ctx->fpt) {
+        free_pool(ctx, true);
+        if (ctx->fph != ctx->fpt)
+            return -1;
+        free(ctx->fpt); /* free the empty node */
+        ctx->fph = ctx->fpt = 0;
+    }
+    if (ctx->fph || ctx->fpt)
+        return -1;
+
+    free(ctx->HP);
+    free(ctx->tid_map);
+    memset(ctx, 0, sizeof(struct lfq_ctx));
+
+    return 0;
+}
+
+int lfq_enqueue(struct lfq_ctx *ctx, void *data)
+{
+    struct lfq_node *insert_node = calloc(1, sizeof(struct lfq_node));
+    if (!insert_node)
+        return -errno;
+
+    insert_node->data = data;
+    struct lfq_node *old_tail = XCHG(&ctx->tail, insert_node);
+    /* We have claimed our spot in the insertion order by modifying tail.
+     * we are the only inserting thread with a pointer to the old tail.
+     *
+     * Now we can make it part of the list by overwriting the NULL pointer in
+     * the old tail. This is safe whether or not other threads have updated
+     * ->next in our insert_node.
+     */
+#ifdef DEBUG
+    assert(!(old_tail->next) && "old tail was not NULL");
+#endif
+    atomic_store(&old_tail->next, insert_node);
+    /* TODO: could a consumer thread could have freed the old tail?  no because
+     * that would leave head=NULL
+     */
+
+    return 0;
+}
+
+void *lfq_dequeue_tid(struct lfq_ctx *ctx, int tid)
+{
+    struct lfq_node *old_head, *new_head;
+
+    /* HP[tid] is necessary for deallocation. */
+    do {
+    retry:
+        /* continue jumps to the bottom of the loop, and would attempt a CAS
+         * with uninitialized new_head.
+         */
+        old_head = atomic_load(&ctx->head);
+
+        /* seq-cst store.
+         * FIXME: use xchg instead of mov + mfence on x86.
+         */
+        atomic_store(&ctx->HP[tid], old_head);
+        mb();
+
+        /* another thread freed it before seeing our HP[tid] store */
+        if (old_head != atomic_load(&ctx->head))
+            goto retry;
+        new_head = atomic_load(&old_head->next);
+
+        if (new_head == 0) {
+            atomic_store(&ctx->HP[tid], 0);
+            return NULL; /* never remove the last node */
+        }
+#ifdef DEBUG
+        // FIXME: check for already-freed nodes
+        // assert(new_head != (void *) -1 && "read an already-freed node");
```

Comment:
Is the assertion valid? If not, remove it.

Reply:
It is not valid. Removed.